File: scripts/generate-dmg-background.mjs
/**
 * generate-dmg-background.mjs
 *
 * Generates DMG background PNGs (1x and 2x) using sharp + SVG.
 * No Python or system dependencies — only sharp, which is already
 * a project dependency.
 *
 * Layout (1x):
 *   Window:  660 × 480
 *   Logo:    560 × 261 px (structura-full-logo.svg, aspect 900:420), centered, top=24
 *   Sep:     y=299
 *   Icons:   APP_X=165 APP_Y=400  FOL_X=495 FOL_Y=400
 *   Arrow:   between the two icon positions
 *   Labels:  none — Finder renders its own labels below each icon
 *
 * Outputs:
 *   src-tauri/icons/dmg-background.png      (660×480, 1x)
 *   src-tauri/icons/dmg-background@2x.png   (1320×960, 2x Retina)
 */

import sharp from "sharp";
import { readFileSync, writeFileSync } from "fs";
import path from "path";
import { fileURLToPath } from "url";

const ROOT = path.dirname(path.dirname(fileURLToPath(import.meta.url)));

const BG     = "#f8f6f2";
const ACCENT = "#c89b3c";
const BORDER = "#ddd8ce";

// Window dimensions — must match create-dmg.mjs window.size
export const W = 660, H = 480;

// Icon center positions — must match create-dmg.mjs contents[].x/y
export const APP_X = 165, APP_Y = 400;
export const FOL_X = 495, FOL_Y = 400;

// Logo dimensions at 1x (SVG viewBox 900×420, rendered at width 560)
const LOGO_W   = 560;
const LOGO_H   = Math.round(LOGO_W * 420 / 900); // ≈ 261
const LOGO_TOP = 24;
const LOGO_LEFT = Math.round((W - LOGO_W) / 2);  // 50

const SEP_Y  = LOGO_TOP + LOGO_H + 14;           // ≈ 299
const ARROW_Y  = APP_Y - 6;
const ARROW_X1 = APP_X + 52;
const ARROW_X2 = FOL_X - 52;
const HEAD     = 10;

async function render(scale) {
  const w   = W * scale, h = H * scale;
  const sw  = Math.max(1, scale);

  const ay  = ARROW_Y   * scale;
  const ax1 = ARROW_X1  * scale, ax2 = ARROW_X2 * scale;
  const hd  = HEAD      * scale;
  const sy  = SEP_Y     * scale;

  const arrowHead = [
    `M ${ax2} ${ay} L ${ax2 - hd} ${ay - hd * 0.6}`,
    `M ${ax2} ${ay} L ${ax2 - hd} ${ay + hd * 0.6}`,
  ].join(" ");

  const svg = `<svg xmlns="http://www.w3.org/2000/svg" width="${w}" height="${h}">
    <rect width="${w}" height="${h}" fill="${BG}"/>
    <line x1="${40 * scale}" y1="${sy}" x2="${(W - 40) * scale}" y2="${sy}"
          stroke="${BORDER}" stroke-width="${sw}"/>
    <line x1="${ax1}" y1="${ay}" x2="${ax2}" y2="${ay}"
          stroke="${ACCENT}" stroke-width="${sw * 2}" stroke-opacity="0.8"/>
    <path d="${arrowHead}" stroke="${ACCENT}" stroke-width="${sw * 2}" stroke-opacity="0.8"
          fill="none" stroke-linecap="round" stroke-linejoin="round"/>
  </svg>`;

  // Rasterise the full Structura SVG logo at the target size
  const logoSvg = readFileSync(path.join(ROOT, "public/structura-full-logo.svg"));
  const logoBuf = await sharp(logoSvg, { density: 300 })
    .resize(LOGO_W * scale, LOGO_H * scale, { fit: "fill" })
    .png()
    .toBuffer();

  return sharp(Buffer.from(svg))
    .png()
    .composite([{
      input: logoBuf,
      left: LOGO_LEFT * scale,
      top:  LOGO_TOP  * scale,
    }])
    .png({ compressionLevel: 9 })
    .toBuffer();
}

console.log("Generating DMG background images…");
for (const [scale, name] of [[1, "dmg-background.png"], [2, "dmg-background@2x.png"]]) {
  const buf = await render(scale);
  const out = path.join(ROOT, "src-tauri/icons", name);
  writeFileSync(out, buf);
  console.log(`  Wrote ${name}  (${W * scale}x${H * scale})`);
}
console.log("Done.");

File: scripts/generate-dmg-background.py
#!/usr/bin/env python3
"""Generate DMG background images for the Structura macOS installer."""

import math
import sys
from pathlib import Path

try:
    from PIL import Image, ImageDraw, ImageFont
except ImportError:
    print("Pillow not found. Install with: pip3 install Pillow")
    sys.exit(1)

# --- Config ---
BG_COLOR = "#f8f6f2"           # matches --background in globals.css
ACCENT_COLOR = "#c89b3c"       # matches --accent
NAV_COLOR = "#1f2f3f"          # matches --foreground / nav-bg
MUTED_COLOR = "#ddd8ce"        # matches --border

W, H = 660, 420                # 1x DMG window size
APP_X, APP_Y = 165, 285        # app icon center (matches tauri.conf.json)
FOLDER_X, FOLDER_Y = 495, 285  # Applications alias center

ICON_PATH = Path(__file__).parent.parent / "src-tauri" / "icons" / "icon.png"
OUT_1X = Path(__file__).parent.parent / "src-tauri" / "icons" / "dmg-background.png"
OUT_2X = Path(__file__).parent.parent / "src-tauri" / "icons" / "dmg-background@2x.png"


def draw_arrow(draw, x1, y1, x2, y2, color, width=2, head=10):
    """Draw a simple horizontal arrow."""
    draw.line([(x1, y1), (x2, y2)], fill=color, width=width)
    # arrowhead
    angle = math.atan2(y2 - y1, x2 - x1)
    for side in [math.pi / 6, -math.pi / 6]:
        ax = x2 - head * math.cos(angle - side)
        ay = y2 - head * math.sin(angle - side)
        draw.line([(x2, y2), (ax, ay)], fill=color, width=width)


def render(scale: int) -> Image.Image:
    w, h = W * scale, H * scale
    img = Image.new("RGBA", (w, h), BG_COLOR)
    draw = ImageDraw.Draw(img)

    # --- Structura icon centered at top ---
    icon_size = 120 * scale
    icon_y_top = 24 * scale          # top of icon
    icon_x_left = (w - icon_size) // 2

    icon = Image.open(ICON_PATH).convert("RGBA")
    icon = icon.resize((icon_size, icon_size), Image.LANCZOS)
    img.paste(icon, (icon_x_left, icon_y_top), icon)

    # --- Thin separator line ---
    sep_y = icon_y_top + icon_size + 18 * scale
    draw.line([(40 * scale, sep_y), (w - 40 * scale, sep_y)],
              fill=MUTED_COLOR, width=max(1, scale))

    # --- "Drag to install" arrow between the two icon positions ---
    arrow_y = APP_Y * scale - 6 * scale      # slightly above icon centers
    arrow_x1 = APP_X * scale + 52 * scale   # right edge of app icon area
    arrow_x2 = FOLDER_X * scale - 52 * scale  # left edge of Applications area

    arrow_color_rgb = tuple(int(ACCENT_COLOR.lstrip("#")[i:i+2], 16) for i in (0, 2, 4))
    arrow_color = arrow_color_rgb + (200,)

    draw_arrow(draw, arrow_x1, arrow_y, arrow_x2, arrow_y,
               color=arrow_color, width=2 * scale, head=10 * scale)

    # --- Labels below the icon positions ---
    label_y = (APP_Y + 46) * scale
    try:
        font_size = 11 * scale
        font = ImageFont.truetype("/System/Library/Fonts/Helvetica.ttc", font_size)
    except Exception:
        font = ImageFont.load_default()

    text_color_rgb = tuple(int(NAV_COLOR.lstrip("#")[i:i+2], 16) for i in (0, 2, 4))
    text_color = text_color_rgb + (160,)

    for label, cx in [("Structura", APP_X * scale), ("Applications", FOLDER_X * scale)]:
        bbox = draw.textbbox((0, 0), label, font=font)
        tw = bbox[2] - bbox[0]
        draw.text((cx - tw // 2, label_y), label, fill=text_color, font=font)

    return img.convert("RGB")


def main():
    print("Generating DMG background images…")
    for scale, out in [(1, OUT_1X), (2, OUT_2X)]:
        img = render(scale)
        img.save(out, "PNG", optimize=True)
        print(f"  Wrote {out}  ({img.width}×{img.height})")
    print("Done.")


if __name__ == "__main__":
    main()
